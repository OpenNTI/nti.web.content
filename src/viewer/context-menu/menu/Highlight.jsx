import React from 'react';
import PropTypes from 'prop-types';

const COLOR_OPTIONS = [
	{name: 'yellow', color: 'EDE619'},
	{name: 'green', color: '4CE67F'},
	{name: 'blue', color: '3FB3F6'}
];

export default class ContextMenuHighlights extends React.Component {
	static propTypes = {
		annotations: PropTypes.bool
	}


	selectOption = (e) => {
		debugger;
	}


	render () {
		const {annotations} = this.props;

		if (!annotations) { return null; }

		return (
			<div className="context-menu-highlights">
				<ul className="options">
					{COLOR_OPTIONS.map((option, key) => {
						const {name, color} = option;

						return (
							<li
								key={key}
								data-option-name={name}
								onClick={this.selectOption}
							>
								<div className="color" style={{backgroundColor: `#${color}`}} />
							</li>
						);
					})}
				</ul>
			</div>
		);
	}
}
