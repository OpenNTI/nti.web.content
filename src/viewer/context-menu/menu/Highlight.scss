@import '~@nti/style-common';

.context-menu-highlights {
	.options {
		list-style: none;
		padding: 0;
		margin: 0;
		display: flex;
		flex-direction: row;
		align-items: center;
		justify-content: flex-start;

		li {
			flex: 0 0 auto;
			padding: 0.5rem 0.2rem;
			cursor: pointer;

			&:first-of-type {
				padding-left: 0.5rem;
			}

			&:last-of-type {
				padding-right: 0.5rem;
			}

			.no-touch &:hover, &.selected {
				.color {
					box-shadow: 0 0 0 2px rgba(255, 255, 255, 0.8);

					&::before {
						content: $icon-check;
						display: inline-block;
						color: $secondary-grey;
						font-size: 0.75rem;
						font-weight: 600;
					}
				}
			}

			.color {
				display: flex;
				justify-content: center;
				align-items: center;
				width: 1.25rem;
				height: 1.25rem;
				border-radius: 50%;
				box-shadow: inset 0 0 0 1px rgba(0, 0, 0, 0.2);

				.nti-content-context-menu.touch & {
					width: 1.875rem;
					height: 1.875rem;
				}
			}

		}
	}
}
