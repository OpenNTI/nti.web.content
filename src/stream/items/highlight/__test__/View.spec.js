import React from 'react';
import renderer from 'react-test-renderer';
import { TestUtils } from '@nti/web-client';

import Highlight from '../View';

const { tearDownTestClient, setupTestClient } = TestUtils;

const mockService = {
	resolveEntity: () => Promise.resolve({ displayName: 'Tim Jones' }),
};

const onBefore = () => {
	jest.useFakeTimers();
	setupTestClient(mockService);
};

const onAfter = () => {
	tearDownTestClient();
};

const flushPromises = () => new Promise(resolve => setImmediate(resolve));

function createNodeMock (element) {
	if (element.type === 'span') {
		return {
			textContext: 'Wide-body aircraft, such as the B-767, B-777, and A-340, have two, three, or four jet engines and may carry over 64,000 gallons (256 000 L) of jet fuel (Figures 3.2a). Wide-body aircraft cabins...',
			setAttribute: () => {},
			querySelector: () => null,
			parentNode: {
				scrollHeight: 0,
				clientHeight: 0,
				offsetHeight: 41
			}
		};
	}
	return null;
}

/* eslint-env jest */
describe('Stream: Highlight', () => {
	beforeEach(onBefore);
	afterEach(onAfter);

	test('General Snapshot', async () => {
		const item = {
			'creator': 'josh.birdwell@nextthought.com',
			'CreatedTime': 1539193071.268633,
			getCreatedTime: function () {
				return new Date(this.CreatedTime);
			},
			getContextPath: function () {
				return Promise.resolve([
					[
						{ NTIID: 'tag:nextthought.com,2011-10:IFSTA-Bundle-IFSTA_Book_Aircraft_Rescue_and_Fire_Fighting_Sixth_Edition', getPresentationProperties: function () { return { author: 'IFSTA', title: 'Aircraft Rescue and Fire Fighting Sixth Edition', label: undefined }; }},
						{ NTIID: 'tag:nextthought.com,2011-10:IFSTA-HTML-IFSTA_Book_Aircraft_Rescue_and_Fire_Fighting_Sixth_Edition.section:Types_of_Aircraft3', Title: 'Types of Aircraft' }
					]
				]);
			},
			'NTIID': 'tag:nextthought.com,2011-10:josh.birdwell@nextthought.com-OID-0x05a7b0:5573657273:kUvXvKw8HbW',
			'applicableRange': {
				'MimeType': 'application/vnd.nextthought.contentrange.domcontentrangedescription',
				'Class': 'DomContentRangeDescription',
				'ancestor': {
					'MimeType': 'application/vnd.nextthought.contentrange.elementdomcontentpointer',
					'Class': 'ElementDomContentPointer',
					'role': 'ancestor',
					'elementTagName': 'P',
					'elementId': 'cb6ffcc62c56c2c4f39b119a9fdd28a8'
				},
				'end': {
					'MimeType': 'application/vnd.nextthought.contentrange.textdomcontentpointer',
					'Class': 'TextDomContentPointer',
					'role': 'end',
					'ancestor': {
						'MimeType': 'application/vnd.nextthought.contentrange.elementdomcontentpointer',
						'Class': 'ElementDomContentPointer',
						'role': 'ancestor',
						'elementTagName': 'P',
						'elementId': 'cb6ffcc62c56c2c4f39b119a9fdd28a8'
					},
					'contexts': [
						{
							'MimeType': 'application/vnd.nextthought.contentrange.textcontext',
							'contextOffset': 121,
							'contextText': 'passengers '
						},
						{
							'MimeType': 'application/vnd.nextthought.contentrange.textcontext',
							'contextOffset': 0,
							'contextText': '('
						},
						{
							'MimeType': 'application/vnd.nextthought.contentrange.textcontext',
							'contextOffset': 0,
							'contextText': 'Figures'
						},
						{
							'MimeType': 'application/vnd.nextthought.contentrange.textcontext',
							'contextOffset': 0,
							'contextText': ')'
						},
						{
							'MimeType': 'application/vnd.nextthought.contentrange.textcontext',
							'contextOffset': 0,
							'contextText': '.'
						}
					],
					'edgeOffset': 10
				},
				'start': {
					'MimeType': 'application/vnd.nextthought.contentrange.textdomcontentpointer',
					'Class': 'TextDomContentPointer',
					'role': 'start',
					'ancestor': {
						'MimeType': 'application/vnd.nextthought.contentrange.elementdomcontentpointer',
						'Class': 'ElementDomContentPointer',
						'role': 'ancestor',
						'elementTagName': 'P',
						'elementId': 'cb6ffcc62c56c2c4f39b119a9fdd28a8'
					},
					'contexts': [
						{
							'MimeType': 'application/vnd.nextthought.contentrange.textcontext',
							'contextOffset': 152,
							'contextText': 'Wide-body'
						}
					],
					'edgeOffset': 0
				},
				'isDomContentRangeDescription': true
			},
			'presentationProperties': {
				'highlightColorName': 'blue'
			},
			selectedText: 'Wide-body aircraft, such as the B-767, B-777, and A-340, have two, three, or four jet engines and may carry over 64,000 gallons (256 000 L) of jet fuel (Figures 3.2a). Wide-body aircraft cabins have dual aisles creating a center section of seats, allowing the aircraft to carry over 500 passengers'
		};
		const context = {
			NTIID: 'tag:nextthought.com,2011-10:IFSTA-Bundle-IFSTA_Book_Aircraft_Rescue_and_Fire_Fighting_Sixth_Edition',
			PlatformPresentationResources: [{
				'Class': 'DisplayablePlatformPresentationResources',
				'CreatedTime': 1532453102,
				'InheritPlatformName': null,
				'Last Modified': 1527882586,
				'PlatformName': 'webapp',
				'Version': 1,
				'href': '/content/sites/ifsta.nextthought.com/ContentPackageBundles/Aircraft%20Rescue%20and%20Fire%20Fighting%20Sixth%20Edition/presentation-assets/webapp/v1/'
			}]
		};

		const options = { createNodeMock };
		const highlightCmp = renderer.create(<Highlight item={item} context={context} />, options);

		jest.runAllTimers();
		await flushPromises();

		const tree = highlightCmp.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
