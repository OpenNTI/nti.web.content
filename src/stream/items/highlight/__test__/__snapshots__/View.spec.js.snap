// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stream: Highlight General Snapshot 1`] = `
<div
  className="stream-highlight"
>
  <div
    className="heading"
  >
    <div
      className="nti-link-to-path"
      to="#"
    >
      <a
        className="username"
        rel="author"
      >
        Tim Jones
      </a>
    </div>
     created a highlight on 
    <time
      dateTime="1970-01-18T13:33:13-06:00"
    >
      January 18, 1970
    </time>
  </div>
  <div
    className="highlight-container"
  >
    <div
      className="nti-link-to-path"
      to="/object/IFSTA-Bundle-IFSTA_Book_Aircraft_Rescue_and_Fire_Fighting_Sixth_Edition"
    >
      <img
        className="highlight-icon"
        src="/content/sites/ifsta.nextthought.com/ContentPackageBundles/Aircraft%20Rescue%20and%20Fire%20Fighting%20Sixth%20Edition/presentation-assets/webapp/v1/contentpackage-thumb-60x60.png?t=1527882586"
      />
    </div>
    <div
      className="highlight-content"
    >
      <div
        className="nti-content-breadcrumb"
      >
        <div
          className="nti-link-to-path"
          to="/object/josh.birdwell%40nextthought.com-OID-0x05a7b0%3A5573657273%3AkUvXvKw8HbW"
        >
          <ul
            className="breadcrumb-list"
          >
            <li
              className="crumb last"
            >
              <span>
                Types of Aircraft
              </span>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="nti-link-to-path"
        to="/object/josh.birdwell%40nextthought.com-OID-0x05a7b0%3A5573657273%3AkUvXvKw8HbW"
      >
        <div
          className="body"
        >
          <span
            className="application-highlight plain blue colored"
            dangerouslySetInnerHTML={
              Object {
                "__html": "Wide-body aircraft, such as the B-767, B-777, and A-340, have two, three, or four jet engines and may carry over 64,000 gallons (256 000 L) of jet fuel (Figures 3.2a). Wide-body aircraft cabins...",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
